### PR TITLE
[server] Allow setting different basepath and static web #1104

### PR DESCRIPTION
**Pull request by pinghe**
Allow setting different basepath. And support for static web services.

### CODE REVIEW

**Comment by michaelvlach** [comment]
Hi, thanks for the PR. What is the use case here? I think I can see the base path being potentially useful but I do not understand why do you want to move the open API schéma to hardcoded "watom" path. Can you please elaborate?

File: agdb_server/src/app.rs [hunk 1]
```diff
@@ -20,7 +21,7 @@ pub(crate) fn app(
 ) -> Router {
     let state = ServerState {
         db_pool,
-        config,
+        config: Arc::clone(&config),
```
Comment (michaelvlach) [inline agdb_server/src/app.rs]: Why not `config.clone()`?

File: agdb_server/src/config.rs [hunk 2]
```diff
@@ -35,6 +36,7 @@ pub(crate) fn new() -> ServerResult<Config> {
     let config = ConfigImpl {
         bind: ":::3000".to_string(),
         address: Url::parse("localhost:3000")?,
+        basepath: "/public".to_string(),
```
Comment (michaelvlach) [inline agdb_server/src/config.rs]: I believe this should be an empty string by default. It will also fix all the tests I think.

**Comment by michaelvlach** [comment]
You should add at least one test, probably in `misc_routes.rs` to verify the base path indeed works. You can get inspiration how to write it from the other tests, particularly those that set something special in the config like `cluster_test.rs`. It is basically just a hashmap you pass to the `TestServerImpl`:

```
let mut config =HashMap::<&str, serde_yaml::Value>::new();
config.insert("basePath", "public".into());
let server = TestServerImpl::with_config(config).await?;
```

**Review by agnesoft — DISMISSED** [review]

**Review by agnesoft — APPROVED** [review]

**Pull request merged**